keps/sig-multicluster/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,911 @@
+*   Orchestrating pod movement from one StatefulSet slice to another
+*   Managing network connectivity between pods in different StatefulSet slices
+*   Orchestrating storage lifecycle of PVCs and PVs across different StatefulSet slices
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This KEP solves the problem of managing subsets of the replicas in a StatefulSet by introducing the concept of a slice. A slice consists of a start ordinal `k`, and a number of replicas `N`. To control the starting and ending ordinal of each slice, a new field `replicaStartOrdinal` is introduced to `StatefulSetSpec`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+The main motivation of this KEP is to support a more flexible StatefulSet, a building block in an ecosystem where Stateful applications can be migrated across Kubernetes clusters with more automation. Below are two high level user stories that share the problem of having a StatefulSet locked into a specific configuration. To fully automate these two scenarios, additional building blocks are needed around volume management and networking (see [Notes/Constraints/Caveats](#notesconstraintscaveats-optional)).
+
+#### Story 1
+
+**Migrating across namespaces**: Many organizations use namespaces for team isolation. Consider a team that is migrating a `StatefulSet` to a new namespace in a cluster. Migration could be motivated by a branding change, or a requirement to move out of a shared namespace. Consider the StatefulSet `my-app` with `replicas: 5`, running in a shared namespace.
+
+
+```
+name: my-app
+namespace: shared
+replicas: 5
+-----------------------------------------------
+[ nginx-0, nginx-1, nginx-2, nginx-3, nginx-4 ]
+```
+
+
+To move two pods, the `my-app` StatefulSet in the `shared` namespace can be scaled down to `replicas: 3, replicaStartOrdinal: 0`, and an analogous StatefulSet in the `app-team` namespace scaled up to `replicas: 2, replicaStartOrdinal: 3`. This allows for pod ordinals to be managed during migration. The application operator should manage network connectivity, volumes and slice orchestration (when to migrate and by how many replicas).
```

Comment:
I think I'm missing why the ordinal is the problem in this use case.  Are you trying to make sure that the workload within the namespace can only look at the pod name to determine what slice they are?  There seem to be some assumptions about how the workload uses the ordinal that would be better to make explicit.

My naive (not understanding the context) answer to this problem would be: "the pod name is already namespaced, so if your workload is using ordinal to separate the workload you should really be using `namespace + pod name + ordinal`" - i.e. that the workload is making an unsafe assumption that should be addressed.

In general, we don't support *any* cross namespace reference on workload, so any details you can add here to explain how the ordinal is used would help.

Reply:
> In general, we don't support any cross namespace reference on workload, so any details you can add here to explain how the ordinal is used would help.
This makes sense, and under steady state conditions I would agree. The problem having a workload split cross namespace or cluster is for re-organization, moving it to a new ownership paradigm. The only supported way to handle this today is re-creation of a workload in a new namespace/cluster, which results in workload downtime.

> My naive (not understanding the context) answer to this problem would be: "the pod name is already namespaced, so if your workload is using ordinal to separate the workload you should really be using namespace + pod name + ordinal" - i.e. that the workload is making an unsafe assumption that should be addressed.
The practical use is to have ordinal to separate the workload across organization units (cluster/namespace), but not to separate the workload logically.

> that the workload is making an unsafe assumption that should be addressed.
What is the unsafe assumption? My takeaway is that if the workload is representing the same logic replica ordinal across organization units (namespace/cluster), this isn't an invariant within Apiserver to enforce this. By doing so, the workload can break at-most-one semantics on a StatefulSet, and it would be up to business logic of an application operator to enforce this invariant, rather than API invariants.

Reply:
> The only supported way to handle this today is re-creation of a workload in a new namespace/cluster, which results in workload downtime.

So you are looking to orchestrate cross cluster moves of statefulsets without downtime - are you trying to do this for *all* statefulsets, or just those that have been adapted to support movement?  Some of these assumptions are that context I was looking for, which helps me understand how you intend to use this and how to review it.

Regarding unsafety, someone orchestrating moves where the workload is ordinal aware (the user has encoded assumptions about the ordinal into the state of each instance and may use them arbitrarily) will need to be very careful, and if the goal is that all statefulsets can be orchestrated, there may be assumptions about namespace / cluster factored in that do require the user to modify their workload.  

To be safe, you also have to verify you have observed the deletion of the “high” ordinals on the source cluster, which if I better understood (perhaps there is a doc somewhere with your orchestration design?) might also have other implications about what changes we need.

One final reason I asked the general question - we decided to limit configurability of ordinals originally for simplicity.  Adding more control would open the door for us to ensure other use cases were not limited by “just adding the start” (an example would be topology aware ordinal assignment), but that might impact the your described negatively.

In general re: statefulset movement, it’s not unreasonable to ask the workload author to fit within some constraints, and “deal with overlapping ordinals” might be a better overall approach if we ended up helping the user to avoid coupling too strongly to a single cluster’s representation.  I just want to be sure we explore that.

Reply:
One more comment - is the ordinal important for networking rules?  That was the heart of my question about cross-namespace - if you are attempting to move across clusters or namespaces you are giving up the ability to maintain a network quorum by looking at the service or the pod name via dns, which is one of the signature properties of stateful workloads under statefulsets.  If you’re willing to live without that, then the ordinal seems almost as unimportant (because you can’t magically make pod-0 in ns1 be unaware of the change in pod-7 to ns2).  Unless you are proposing the orchestrator handle that as well?

Knowing what level of intent you have to make this transparent helps refine how critical this is.

Reply:
Thanks, will review your responses in more detail before I respond.  One immediate question:

> Allowing a StatefulSet to start from an arbitrary ordinal might be adding too much configurability. This could also be achieved with a reverse scale-up flag (eg: scale up "N-1" instead of "0", as is the default with OrderedReady pod management today).

Another option might be to put a stateful set in a "paused mode" or to add a way for individual ordinals to be marked as "paused" at the workload object level, then orphan the pods one by one viat he orchestrator.  The ability for an admin to "freeze" a particular ordinal might very well be generally useful for debugging or recovery, and we might be able to find ways to overlap that use case with other operator behavior today.  I know that currently if you wanted to perform debugging on a particular pod, you might want to ensure that upgrades / etc are blocked / held by that pod, or test how the workload works in the absence of that pod.  I don't know how complex that might get, but it would be one way to overlap the ask with something that grows the utility of workloads on a single cluster as well.

Reply:
> Another option might be to put a stateful set in a "paused mode" or to add a way for individual ordinals to be marked as "paused" at the workload object level ...  The ability for an admin to "freeze" a particular ordinal might very well be generally useful for debugging or recovery, and we might be able to find ways to overlap that use case with other operator behavior today
If there is an API to freeze an ordinal level (eg: KEP-3521 or otherwise), the mechanics of pausing a pod ordinal could be very similar to the workflow I described here: https://github.com/kubernetes/enhancements/pull/3522#issuecomment-1251634828.